README.md
# pyzk

[![Build Status](https://travis-ci.org/fananimi/pyzk.svg?branch=master)](https://travis-ci.org/fananimi/pyzk)

`pyzk` is an unofficial library of zksoftware (zkteco family) attendance machine.

# Installation

There is some installation method you can use below:

* pip
```sh
pip install -U pyzk
```

* manual installation (clone and execute)

```sh
python setup.py install
```

* clone and append the path of this project

```python
import sys
import os
sys.path.insert(1,os.path.abspath("./pyzk"))
from zk import ZK, const
```

# Documentation

Complete documentation can be found at [Readthedocs](http://pyzk.readthedocs.io/en/latest/ "pyzk's readthedocs") .

# API Usage

Create the ZK object and you will be ready to call api.

## Basic Usage

The following is an example code block how to use pyzk.

```python
from zk import ZK, const

conn = None
# create ZK instance
zk = ZK('192.168.1.201', port=4370, timeout=5, password=0, force_udp=False, ommit_ping=False)
try:
    # connect to device
    conn = zk.connect()
    # disable device, this method ensures no activity on the device while the process is run
    conn.disable_device()
    # another commands will be here!
    # Example: Get All Users
    users = conn.get_users()
    for user in users:
        privilege = 'User'
        if user.privilege == const.USER_ADMIN:
            privilege = 'Admin'
        print ('+ UID #{}'.format(user.uid))
        print ('  Name       : {}'.format(user.name))
        print ('  Privilege  : {}'.format(privilege))
        print ('  Password   : {}'.format(user.password))
        print ('  Group ID   : {}'.format(user.group_id))
        print ('  User  ID   : {}'.format(user.user_id))

    # Test Voice: Say Thank You
    conn.test_voice()
    # re-enable device after all commands already executed
    conn.enable_device()
except Exception as e:
    print ("Process terminate : {}".format(e))
finally:
    if conn:
        conn.disconnect()
```

## Command List

* Connect/Disconnect

```python
conn = zk.connect()
conn.disconnect()
```

* Disable/Enable Connected Device

```python
# disable (lock) device, to ensure no user activity in device while some process run
conn.disable_device()
# re-enable the connected device and allow user activity in device again
conn.enable_device()
```

* Get and Set Time

```python
from datetime import datetime
# get current machine's time
zktime = conn.get_time()
print zktime
# update new time to machine
newtime = datetime.today()
conn.set_time(newtime)
```


* Ger Firmware Version and Extra Information

```python
conn.get_firmware_version()
conn.get_serialnumber()
conn.get_platform()
conn.get_device_name()
conn.get_face_version()
conn.get_fp_version()
conn.get_extend_fmt()
conn.get_user_extend_fmt()
conn.get_face_fun_on()
conn.get_compat_old_firmware()
conn.get_network_params()
conn.get_mac()
conn.get_pin_width()
```

* Get Device Usage Space

```python
conn.read_sizes()
print(conn)
#also:
conn.users
conn.fingers
conn.records
conn.users_cap
conn.fingers_cap
# TODO: add records_cap counter
# conn.records_cap
```

* User Operation

```python
# Create user
conn.set_user(uid=1, name='Fanani M. Ihsan', privilege=const.USER_ADMIN, password='12345678', group_id='', user_id='123', card=0)
# Get all users (will return list of User object)
users = conn.get_users()
# Delete User
conn.delete_user(uid=1)
conn.delete_user(user_id=123)
```

* Fingerprints

```python
# Get  a single Fingerprint (will return a Finger object)
template = conn.get_user_template(uid=1, temp_id=0) #temp_id is the finger to read 0~9
# Get all fingers from DB (will return a list of Finger objects)
fingers = conn.get_templates()

# to restore a finger, we need to assemble with the corresponding user
# pass a User object and a list of finger (max 10) to save
conn.save_user_template(user, [fing1 ,fing2])
```

* Remote Fingerprint Enrollment
```python
zk.enroll_user('1')
# but it doesn't work with some tcp ZK8 devices
```


* Attendance Record
```python
# Get attendances (will return list of Attendance object)
attendances = conn.get_attendance()
# Clear attendances records
conn.clear_attendance()
```

* Test voice

```python
"""
 play test voice:
  0 Thank You
  1 Incorrect Password
  2 Access Denied
  3 Invalid ID
  4 Please try again
  5 Dupicate ID
  6 The clock is flow
  7 The clock is full
  8 Duplicate finger
  9 Duplicated punch
  10 Beep kuko
  11 Beep siren
  12 -
  13 Beep bell
  14 -
  15 -
  16 -
  17 -
  18 Windows(R) opening sound
  19 -
  20 Fingerprint not emolt
  21 Password not emolt
  22 Badges not emolt
  23 Face not emolt
  24 Beep standard
  25 -
  26 -
  27 -
  28 -
  29 -
  30 Invalid user
  31 Invalid time period
  32 Invalid combination
  33 Illegal Access
  34 Disk space full
  35 Duplicate fingerprint
  36 Fingerprint not registered
  37 -
  38 -
  39 -
  40 -
  41 -
  42 -
  43 -
  43 -
  45 -
  46 -
  47 -
  48 -
  49 -
  50 -
  51 Focus eyes on the green box
  52 -
  53 -
  54 -
  55 -
"""
conn.test_voice(index=0) # will say 'Thank You'
```

* Device Maintenance

```python
# DANGER!!! This command will be erase all data in the device (incuded: user, attendance report, and finger database)
conn.clear_data()
# shutdown connected device
conn.poweroff()
# restart connected device
conn.restart()
# clear buffer
conn.free_data()
```

* Live Capture!

```python
# live capture! (timeout at 10s)
for attendance in conn.live_capture():
    if attendance is None:
        # implement here timeout logic
        pass
    else:
        print (attendance) # Attendance object

    #if you need to break gracefully just set
    #   conn.end_live_capture = True
    #
    # On interactive mode,
    # use Ctrl+C to break gracefully
    # this way it restores timeout
    # and disables live capture
```

**Test Machine**

```sh
usage: ./test_machine.py [-h] [-a ADDRESS] [-p PORT] [-T TIMEOUT] [-P PASSWORD]
                         [-f] [-t] [-r] [-u] [-l] [-D DELETEUSER] [-A ADDUSER]
                         [-E ENROLLUSER] [-F FINGER]

ZK Basic Reading Tests

optional arguments:
  -h, --help            show this help message and exit
  -a ADDRESS, --address ADDRESS
                        ZK device Address [192.168.1.201]
  -p PORT, --port PORT  ZK device port [4370]
  -T TIMEOUT, --timeout TIMEOUT
                        Default [10] seconds (0: disable timeout)
  -P PASSWORD, --password PASSWORD
                        Device code/password
  -b, --basic           get Basic Information only. (no bulk read, ie: users)
  -f, --force-udp       Force UDP communication
  -v, --verbose         Print debug information
  -t, --templates       Get templates / fingers (compare bulk and single read)
  -tr, --templates-raw  Get raw templates (dump templates)
  -r, --records         Get attendance records
  -u, --updatetime      Update Date/Time
  -l, --live-capture    Live Event Capture
  -o, --open-door       Open door

  -D DELETEUSER, --deleteuser DELETEUSER
                        Delete a User (uid)
  -A ADDUSER, --adduser ADDUSER
                        Add a User (uid) (and enroll)
  -E ENROLLUSER, --enrolluser ENROLLUSER
                        Enroll a User (uid)
  -F FINGER, --finger FINGER
                        Finger for enroll (fid=0)

```

**Backup/Restore (Users and fingers only!!!)** *(WARNING! destructive test! do it at your own risk!)*

```sh
usage: ./test_backup_restore.py [-h] [-a ADDRESS] [-p PORT] [-T TIMEOUT]
                              [-P PASSWORD] [-f] [-v] [-r]
                              [filename]

ZK Basic Backup/Restore Tool

positional arguments:
  filename              backup filename (default [serialnumber].bak)

optional arguments:
  -h, --help            show this help message and exit
  -a ADDRESS, --address ADDRESS
                        ZK device Address [192.168.1.201]
  -p PORT, --port PORT  ZK device port [4370]
  -T TIMEOUT, --timeout TIMEOUT
                        Default [10] seconds (0: disable timeout)
  -P PASSWORD, --password PASSWORD
                        Device code/password
  -f, --force-udp       Force UDP communication
  -v, --verbose         Print debug information
  -E, --erase           clean the device after writting backup!
  -r, --restore         Restore from backup
  -c, --clear-attendance
                        On Restore, also clears the attendance [default keep
                        attendance]
```

To restore on a different device, make sure to specify the `filename`. on restoring, it asks for the serial number of the destination device (to make sure it was correct, as it deletes all data) WARNING. there is no way to restore attendance data, you can keep it or clear it, but once cleared, there is no way to restore it. 

# Compatible devices

```
Firmware Version : Ver 6.21 Nov 19 2008
Platform : ZEM500
DeviceName : U580

Firmware Version : Ver 6.60 Apr 9 2010
Platform : ZEM510_TFT
DeviceName : T4-C

Firmware Version : Ver 6.60 Dec 1 2010
Platform : ZEM510_TFT
DeviceName : T4-C

Firmware Version : Ver 6.60 Mar 18 2011
Platform : ZEM600_TFT
DeviceName : iClock260

Platform         : ZEM560_TFT
Firmware Version : Ver 6.60 Feb  4 2012
DeviceName       :

Firmware Version : Ver 6.60 Oct 29 2012
Platform : ZEM800_TFT
DeviceName : iFace402/ID

Firmware Version : Ver 6.60 Mar 18 2013
Platform : ZEM560
DeviceName : MA300

Firmware Version : Ver 6.60 Dec 27 2014
Platform : ZEM600_TFT
DeviceName : iFace800/ID

Firmware Version : Ver 6.60 Nov 6 2017 (remote tested with correct results)
Platform : ZMM220_TFT
DeviceName : (unknown device) (broken info but at least the important data was read)

Firmware Version : Ver 6.60 Jun 9 2017
Platform : JZ4725_TFT
DeviceName : K20 (latest checked correctly!)

Firmware Version : Ver 6.60 Aug 23 2014 
Platform : ZEM600_TFT
DeviceName : VF680 (face device only, but we read the user and attendance list!)

Firmware Version : Ver 6.70 Feb 16 2017
Platform : ZLM30_TFT
DeviceName : RSP10k1 (latest checked correctly!)

Firmware Version : Ver 6.60 Jun 16 2015
Platform : JZ4725_TFT
DeviceName : K14 (tested & verified working as expected.)

Firmware Version : Ver 6.60 Jan 13 2016
Platform         : ZMM220_TFT
DeviceName       : iFace702 (without voice function, test with encoding='gbk')
```



### Latest tested (not really confirmed)

```
Firmware Version : Ver 6.60 Jun 16 2015
Platform : JZ4725_TFT
DeviceName : iClock260

Firmware Version : Ver 6.60 Jun 5 2015
Platform : ZMM200_TFT
DeviceName : iClock3000/ID (Active testing! latest fix)

Firmware Version : Ver 6.70 Jul 12 2013
Platform : ZEM600_TFT
DeviceName : iClock880-H/ID (Active testing! latest fix)
```

### Not Working (needs more tests, more information)

```
Firmware Version : Ver 6.4.1 (build 99) (display version 2012-08-31)
Platform : 
DeviceName : iClock260 (no capture data - probably similar problem as the latest TESTED)
```

If you have another version tested and it worked, please inform me to update this list!

# Todo

* Create better documentation
* ~~Finger template downloader & uploader~~
* HTTP Rest api
* ~~Create real time api (if possible)~~
* and much more ...



CHANGELOG.md
Changelog
=========
Version 0.9
-----------
* Python 3 Support
* major changes

Version 0.8
-----------
* test suite tool (test_machine.py)
* Initial TCP support

Version 0.7
-----------
* internal major changes

Version 0.6
-----------
* device password support

Version 0.5
-----------
* bug fixed get_users bug

Version 0.4
-----------
* bug fixed
* minor update
* update documentation

Version 0.3
-----------
* add function `get_serialnumber` (return device serial number)
* add function `delete_user` (delete specific user by uid)
* add function `clear_data` (format device)
* add function `get_attendance` (get all attendance records)
* add function `clear_attendance` (clear all attendance records)

Version 0.2
------------
* add basic function and program fundamental
* configure pypi
* configure travis-ci integration

Version 0.1
-----------
* project initialization


zk/const.py
# -*- coding: utf-8 -*-

USHRT_MAX = 65535

CMD_DB_RRQ          = 7     # Read in some kind of data from the machine
CMD_USER_WRQ        = 8     # Upload the user information (from PC to terminal).
CMD_USERTEMP_RRQ    = 9     # Read some fingerprint template or some kind of data entirely
CMD_USERTEMP_WRQ    = 10    # Upload some fingerprint template
CMD_OPTIONS_RRQ     = 11    # Read in the machine some configuration parameter
CMD_OPTIONS_WRQ     = 12    # Set machines configuration parameter
CMD_ATTLOG_RRQ      = 13    # Read all attendance record
CMD_CLEAR_DATA      = 14    # clear Data
CMD_CLEAR_ATTLOG    = 15    # Clear attendance records
CMD_DELETE_USER     = 18    # Delete some user
CMD_DELETE_USERTEMP = 19    # Delete some fingerprint template
CMD_CLEAR_ADMIN     = 20    # Cancel the manager
CMD_USERGRP_RRQ     = 21    # Read the user grouping
CMD_USERGRP_WRQ     = 22    # Set users grouping
CMD_USERTZ_RRQ      = 23    # Read the user Time Zone set
CMD_USERTZ_WRQ      = 24    # Write the user Time Zone set
CMD_GRPTZ_RRQ       = 25    # Read the group Time Zone set
CMD_GRPTZ_WRQ       = 26    # Write the group Time Zone set
CMD_TZ_RRQ          = 27    # Read Time Zone set
CMD_TZ_WRQ          = 28    # Write the Time Zone
CMD_ULG_RRQ         = 29    # Read unlocks combination
CMD_ULG_WRQ         = 30    # write unlocks combination
CMD_UNLOCK          = 31    # unlock
CMD_CLEAR_ACC       = 32    # Restores Access Control set to the default condition.
CMD_CLEAR_OPLOG     = 33    # Delete attendance machines all attendance record.
CMD_OPLOG_RRQ       = 34    # Read manages the record
CMD_GET_FREE_SIZES  = 50    # Obtain machines condition, like user recording number and so on
CMD_ENABLE_CLOCK    = 57    # Ensure the machine to be at the normal work condition
CMD_STARTVERIFY     = 60    # Ensure the machine to be at the authentication condition
CMD_STARTENROLL     = 61    # Start to enroll some user, ensure the machine to be at the registration user condition
CMD_CANCELCAPTURE   = 62    # Make the machine to be at the waiting order status, please refers to the CMD_STARTENROLL description.
CMD_STATE_RRQ       = 64    # Gain the machine the condition
CMD_WRITE_LCD       = 66    # Write LCD
CMD_CLEAR_LCD       = 67    # Clear the LCD captions (clear screen).
CMD_GET_PINWIDTH    = 69    # Obtain the length of user’s serial number
CMD_SMS_WRQ         = 70    # Upload the short message.
CMD_SMS_RRQ         = 71    # Download the short message
CMD_DELETE_SMS      = 72    # Delete the short message
CMD_UDATA_WRQ       = 73    # Set user’s short message
CMD_DELETE_UDATA    = 74    # Delete user’s short message
CMD_DOORSTATE_RRQ   = 75    # Obtain the door condition
CMD_WRITE_MIFARE    = 76    # Write the Mifare card
CMD_EMPTY_MIFARE    = 78    # Clear the Mifare card

CMD_GET_TIME        = 201   # Obtain the machine time
CMD_SET_TIME        = 202   # Set machines time
CMD_REG_EVENT       = 500   # Register the event

CMD_CONNECT         = 1000  # Connections requests
CMD_EXIT            = 1001  # Disconnection requests
CMD_ENABLEDEVICE    = 1002  # Ensure the machine to be at the normal work condition
CMD_DISABLEDEVICE   = 1003  # Make the machine to be at the shut-down condition, generally demonstrates ‘in the work ...’on LCD
CMD_RESTART         = 1004  # Restart the machine.
CMD_POWEROFF        = 1005  # Shut-down power source
CMD_SLEEP           = 1006  # Ensure the machine to be at the idle state.
CMD_RESUME          = 1007  # Awakens the sleep machine (temporarily not to support)
CMD_CAPTUREFINGER   = 1009  # Captures fingerprints picture
CMD_TEST_TEMP       = 1011  # Test some fingerprint exists or does not
CMD_CAPTUREIMAGE    = 1012  # Capture the entire image
CMD_REFRESHDATA     = 1013  # Refresh the machine interior data
CMD_REFRESHOPTION   = 1014  # Refresh the configuration parameter
CMD_TESTVOICE       = 1017  # Play voice
CMD_GET_VERSION     = 1100  # Obtain the firmware edition
CMD_CHANGE_SPEED    = 1101  # Change transmission speed
CMD_AUTH            = 1102  # Connections authorizations
CMD_PREPARE_DATA    = 1500  # Prepares to transmit the data
CMD_DATA            = 1501  # Transmit a data packet
CMD_FREE_DATA       = 1502  # Clear machines opened buffer

CMD_ACK_OK          = 2000  # Return value for order perform successfully
CMD_ACK_ERROR       = 2001  # Return value for order perform failed
CMD_ACK_DATA        = 2002  # Return data
CMD_ACK_RETRY       = 2003  # * Regstered event occorred */
CMD_ACK_REPEAT      = 2004  # Not available
CMD_ACK_UNAUTH      = 2005  # Connection unauthorized

CMD_ACK_UNKNOWN     = 0xffff# Unkown order
CMD_ACK_ERROR_CMD   = 0xfffd# Order false
CMD_ACK_ERROR_INIT  = 0xfffc#/* Not Initializated */
CMD_ACK_ERROR_DATA  = 0xfffb# Not available

EF_ATTLOG           = 1     # Be real-time to verify successfully
EF_FINGER           = (1<<1)# be real–time to press fingerprint (be real time to return data type sign)
EF_ENROLLUSER       = (1<<2)# Be real-time to enroll user
EF_ENROLLFINGER     = (1<<3)# be real-time to enroll fingerprint
EF_BUTTON           = (1<<4)# be real-time to press button
EF_UNLOCK           = (1<<5)# be real-time to unlock
EF_VERIFY           = (1<<7)# be real-time to verify fingerprint
EF_FPFTR            = (1<<8)# be real-time capture fingerprint minutia
EF_ALARM            = (1<<9)# Alarm signal

USER_DEFAULT        = 0
USER_ENROLLER       = 2
USER_MANAGER        = 6
USER_ADMIN          = 14

FCT_ATTLOG          = 1
FCT_WORKCODE        = 8
FCT_FINGERTMP       = 2
FCT_OPLOG           = 4
FCT_USER            = 5
FCT_SMS             = 6
FCT_UDATA           = 7

MACHINE_PREPARE_DATA_1 = 20560 # 0x5050
MACHINE_PREPARE_DATA_2 = 32130 # 0x7282